src/serac/physics/functional_weak_form.hpp

```diff
@@ -0,0 +1,413 @@
+   * integrand involving 'right hand side' contributions like a body load, should be supplied by the user with a
+   * negative sign.
+   * @pre integrand must be a object that can be called with the following arguments:
+   *    1. `double t` the time
+   *    2. `tuple{tensor<T,dim>, isoparametric derivative} X` the spatial coordinates for the quadrature point and the
+   * coordinate's isoparametric derivative.
+   *    3. `tuple{value, derivative}`, a variadic list of tuples (each with a values and spatial derivative),
+   *            one tuple for each of the trial spaces specified in the `DependsOn<...>` argument.
+   * @note The actual types of these arguments passed will be `double`, `tensor<double, ... >` or tuples thereof
+   *    when doing direct evaluation. When differentiating with respect to one of the inputs, its stored
+   *    values will change to `dual` numbers rather than `double`. (e.g. `tensor<double,3>` becomes `tensor<dual<...>,
+   * 3>`)
+   *
+   */
+  template <int... active_parameters, typename BodyIntegralType>
+  void addBodyIntegral(DependsOn<active_parameters...>, std::string body_name, BodyIntegralType integrand)
+  {
+    weak_form_->AddDomainIntegral(Dimension<spatial_dim>{}, DependsOn<active_parameters...>{}, integrand,
+                                  mesh_->domain(body_name));
+
+    v_dot_weak_form_residual_->AddDomainIntegral(
+        Dimension<spatial_dim>{}, DependsOn<0, 1 + active_parameters...>{},
+        [integrand](double t, auto X, auto V, auto... inputs) {
+          auto orig_tuple = integrand(t, X, inputs...);
+          return serac::inner(get<VALUE>(V), get<VALUE>(orig_tuple)) +
+                 serac::inner(get<DERIVATIVE>(V), get<DERIVATIVE>(orig_tuple));
+        },
+        mesh_->domain(body_name));
+  }
+
+  /// @overload
+  template <typename BodyForceType>
+  void addBodyIntegral(std::string body_name, BodyForceType body_integral)
+  {
+    addBodyIntegral(DependsOn<>{}, body_name, body_integral);
+  }
+
+  /**
+   * @brief Add a body source (body load) to the weak form
+   *
+   * @tparam active_parameters Type for indices into fields which the body integral may depend on
+   * @tparam BodyLoadType The type of the body load function
+   * @param body_name The name of the registered domain over which the body loads are applied.
+   * @param depends_on Andices into fields which the body integral may depend on
+   * @param load_function A function describing the body force applied.
+   * @pre load_function must be a object that can be called with the following arguments:
+   *    1. `double t` the time
+   *    2. `tensor<T,dim> X` the spatial coordinates for the quadrature point.
+   *    3. `value`, a variadic list of field values, one tuple for each of the trial spaces specified in the
+   * `DependsOn<...>` argument.
+   *    The expected return is the value (either double or tensor<double,field_dim>) of the body source/load flux
```

Comment:
```suggestion
   *    The expected return is the value of the source at X.
```

I removed the output shape requirement because the either/or description was vague about the circumstances. I started to clarify it, saying that it had to have the same tensor shape as the test function. However, I think this is worse. For people who know what a source is, it's redundant, and for those who don't, it's not enough to help.